pywhy_stats/independence/kci.py

```diff
@@ -153,40 +204,19 @@ def _kernel_test(
     null_sample_size: int,
     threshold: float,
     normalize_data: bool,
+    centered: bool,
+    n_jobs: Optional[int],
 ) -> Tuple[float, float]:
-    if X.ndim == 1:
-        X = X.reshape(-1, 1)
-    if Y.ndim == 1:
-        Y = Y.reshape(-1, 1)
-    if Z is not None and Z.ndim == 1:
-        Z = Z.reshape(-1, 1)
-
-    if normalize_data:
-        # first normalize the data to have zero mean and unit variance
-        # along the columns of the data
-        X = _normalize_data(X)
-        Y = _normalize_data(Y)
-        if Z is not None:
-            Z = _normalize_data(Z)
-
-    if kernel_X is None:
-        kernel_X = _get_default_kernel(X)
-    if kernel_Y is None:
-        kernel_Y = _get_default_kernel(Y)
-
-    Kx = kernel_X(X)
-    Ky = kernel_Y(Y)
+    X, Y, Z = _preprocess_kernel_data(X, Y, Z, normalize_data)
+
+    # compute kernels in each data space
     if Z is not None:
-        if kernel_Z is None:
-            kernel_Z = _get_default_kernel(Z)
-        Kz = kernel_Z(Z)
-        # Equivalent to concatenating them beforehand.
-        # However, here we can then have individual kernels.
-        Kx = Kx * Kz
-        Kz = _fast_centering(Kz)
+        Kz = compute_kernel(Z, metric=kernel_Z, centered=centered, n_jobs=n_jobs)
 
-    Kx = _fast_centering(Kx)
-    Ky = _fast_centering(Ky)
+        # concatenate the (X, Z) data to compute the K_xz kernel
+        X = np.concatenate((X, Z), axis=1)
```

Comment:
This will cause error with customized kernels and/or mixed data types. For instance, if X is continous and Z is categorical (strings) and a user only defines a RBF kernel (or we use it by default), we cannot apply it to the concatenated version. Therefore, I used a pointwise multiplication of the kernel matrices before (which is equivalent), in that way, we still have our individual kernels for each variable.

Reply:
In this regards, the `kernel_Z` is not used anymore?

Reply:
Another follow-up: Providing these custom kernels does not seem to be compatible with the `compute_kernel` logic. What about a simple logic like:
```
if isinstance(kernel_X, str):
   kernel_X = compute_kernel(X, metric=kernel_X, centered=centered, n_jobs=n_jobs)

Kx = kernel_X(X)
```

then, we can still support strings and use the other metric logic from scikit, but can also allow completely customized kernels. The only thing that becomes slightly more complicated would custom metric implementations, but even there, a user can simply pass: `kernel_X = lambda x: compute_kernel(x, metric=my_custom_metric, centered=centered, n_jobs=n_jobs)` (or use partial).

What do you think? I feel like this is a compromise that allows both.

Reply:
Related: https://github.com/py-why/pywhy-stats/pull/19/files#r1287466282

I think I have to think about this a bit more. 

```
if isinstance(kernel_X, str):
   kernel_X = compute_kernel(X, metric=kernel_X, centered=centered, n_jobs=n_jobs)

Kx = kernel_X(X)
```

inherently assumes tho the `kernel_X` callable is a certain API signature. If we tell users in the doc-string this is how it will be called, will this be sufficient? I am okay w/ this pattern, but this pattern was not sufficient to cover all use-cases I had so hence the refactoring.

But the if/else pattern can then capture both cases. Just need to make sure this is what we **require** in the documentation, otw we just error out and say the callable they passed is incorrect?